File: app/src/main/java/osama/me/viewmodelannotations/MainRecyclerViewAdapter.java
package osama.me.viewmodelannotations;


import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

public class MainRecyclerViewAdapter extends BaseAdapter {

    private List<AKindOfObject> data = new ArrayList<>();

    public MainRecyclerViewAdapter(List<AKindOfObject> data) {
        super();
        this.data = data;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(final ViewGroup parent, final int viewType) {
        return new MainItemViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.layout_item, parent, false));
    }

    @Override
    @SuppressWarnings("unchecked")
    public void onBindViewHolder(final RecyclerView.ViewHolder holder, final int position) {
        ((BaseViewHolder) holder).bind(getObject(position));
    }

    private Object getObject(int index) {
        return data.get(index);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }
}

File: app/src/main/java/osama/me/viewmodelannotations/MainRecyclerViewAdapter.kt
package osama.me.viewmodelannotations

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup

class MainRecyclerViewAdapter(val data: List<AKindOfObject>) : BaseAdapter() {


    override fun onBindViewHolder(holder: BaseViewHolder<AKindOfObject>?, position: Int) {
        holder?.bind(getObject(position) as AKindOfObject)
    }

    override fun getItemCount(): Int {
        return data.size
    }

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): BaseViewHolder<AKindOfObject> {
        return MainItemViewHolder(LayoutInflater.from(parent?.context).inflate(R.layout.layout_item, parent, false))
    }

    private fun getObject(index: Int): Any {
        return data[index]
    }
}


abstract class BaseAdapter : RecyclerView.Adapter<BaseViewHolder<AKindOfObject>>()